### PR TITLE
Implement channel_type negotiation

### PR DESCRIPTION
**Pull request by TheBlueMatt**
This implements channel_type negotiation as defined at https://github.com/lightningnetwork/lightning-rfc/pull/880, though it does not currently send the proposed feature bit from https://github.com/lightningnetwork/lightning-rfc/pull/906. It is likely to be required for turbo channels, but if not will be required for anchor channels anyway.

It could use a test of some form, at least of the fallback behavior.

### CODE REVIEW

File: lightning-background-processor/src/lib.rs [hunk 1]
```diff
@@ -153,7 +153,7 @@ impl BackgroundProcessor {
 	/// functionality implemented by other handlers.
 	/// * [`NetGraphMsgHandler`] if given will update the [`NetworkGraph`] based on payment failures.
 	///
-	/// [top-level documentation]: Self
+	/// [top-level documentation]: BackgroundProcessor
```
Comment (arik-so) [inline lightning-background-processor/src/lib.rs:156]: not really part of this PR, but whatever

File: lightning/src/ln/channel.rs [hunk 2]
```diff
@@ -4204,6 +4204,7 @@ impl<Signer: Sign> Channel<Signer> {
 				Some(script) => script.clone().into_inner(),
 				None => Builder::new().into_script(),
 			}),
+			channel_type: None,
```
Comment (arik-so) [inline lightning/src/ln/channel.rs]: unrelated to this diff, but if the message is to obtain an unprompted message to send, "generate" is probably a better prefix than "get"

File: lightning/src/ln/msgs.rs [hunk 3]
```diff
@@ -2052,7 +2059,7 @@ mod tests {
 		do_encoding_channel_update(true, true, true, true);
 	}
 
-	fn do_encoding_open_channel(random_bit: bool, shutdown: bool) {
+	fn do_encoding_open_channel(random_bit: bool, shutdown: bool, chan_type: bool) {
```
Comment (arik-so) [inline lightning/src/ln/msgs.rs:2062]: channel_type is a bitmap or feature set. The boolean should be named `encode_channel_type` or `should_encode_channel_type` or sth. along those lines.

**Review by arik-so — APPROVED** [review]
Looks good with the exception of minor nits.

File: lightning/src/ln/channel.rs (hunk 2 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs]: Yea, :man_shrugging:, I mean its getting *the* open channel? I'm not gonna bother changing it unless you feel strongly.

File: lightning/src/ln/msgs.rs (hunk 3 above)
Comment (TheBlueMatt) [reply lightning/src/ln/msgs.rs:2062]: went with "incl" ie "include".

File: lightning-background-processor/src/lib.rs (hunk 1 above)
Comment (TheBlueMatt) [reply lightning-background-processor/src/lib.rs:156]: :man_shrugging: not worth a separate pr for it.

File: lightning/src/ln/channel.rs (hunk 2 above)
Comment (arik-so) [reply lightning/src/ln/channel.rs]: Fair point

File: lightning/src/ln/features.rs [hunk 4]
```diff
@@ -442,6 +478,21 @@ impl InvoiceFeatures {
 	}
 }
 
+impl ChannelTypeFeatures {
+	/// Constructs the implicit channel type based on the common supported types between us and our
+	/// counterparty
+	pub(crate) fn from_counterparty_init(counterparty_init: &InitFeatures) -> Self {
+		counterparty_init.to_context_internal()
```
Comment (valentinewallace) [inline lightning/src/ln/features.rs:485]: If they only *support* static_remotekey (i.e., they don't require it), should we translate their features to require static_remotekey?

File: lightning/src/ln/features.rs [hunk 5]
```diff
@@ -553,6 +604,20 @@ impl<T: sealed::Context> Features<T> {
 		&self.flags
 	}
 
+	fn write_be<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		for f in self.flags.iter().rev() { // Swap back to big-endian
+			f.write(w)?;
+		}
+		Ok(())
+	}
```
Comment (valentinewallace) [inline lightning/src/ln/features.rs:612]: nit: newline below

File: lightning/src/ln/channel.rs [hunk 6]
```diff
@@ -5393,15 +5421,26 @@ impl<'a, Signer: Sign, K: Deref> ReadableArgs<&'a K> for Channel<Signer>
 
 		let mut announcement_sigs = None;
 		let mut target_closing_feerate_sats_per_kw = None;
+		// Prior to supporting channel type negotiation, all of our channels were static_remotekey
+		// only, so we default to that if none was written.
+		let mut channel_type = Some(ChannelTypeFeatures::only_static_remote_key());
```
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:5426]: should we do `known()` here instead?

File: lightning/src/ln/features.rs [hunk 7]
```diff
@@ -692,25 +757,35 @@ impl<T: sealed::ShutdownAnySegwit> Features<T> {
...
 }
+impl_feature_len_prefixed_write!(InitFeatures);
+impl_feature_len_prefixed_write!(ChannelFeatures);
+impl_feature_len_prefixed_write!(NodeFeatures);
+impl_feature_len_prefixed_write!(InvoiceFeatures);
 
-impl<T: sealed::Context> Readable for Features<T> {
+impl Writeable for ChannelTypeFeatures {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		self.write_be(w)
```
Comment (valentinewallace) [inline lightning/src/ln/features.rs:782]: Why not just use `impl_feature_len_prefixed_write` for simplicity? If not, comment would be good

File: lightning/src/ln/features.rs (hunk 4 above)
Comment (TheBlueMatt) [reply lightning/src/ln/features.rs:485]: The BOLT says every bit uses the even/required form, so I dunno if we should? We do consider a channel type features object as "dont know it" if it has odd bits, I believe.

https://github.com/lightningnetwork/lightning-rfc/commit/4372f9d5380005a0e860875cf403cc0ee4211759#diff-ed04ca2c673fd6aabde69389511fa9ee60cb44d6b2ef6c88b549ffaa753d6afeR206-R214

File: lightning/src/ln/channel.rs (hunk 6 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:5426]: That's a good point - I wanted to make it explicit, though, as we may have to iterate types in the future.

File: lightning/src/ln/features.rs [hunk 8]
```diff
@@ -388,6 +412,18 @@ pub type ChannelFeatures = Features<sealed::ChannelContext>;
 /// Features used within an invoice.
 pub type InvoiceFeatures = Features<sealed::InvoiceContext>;
 
+/// Features used within the channel_type field in an OpenChannel message.
+///
+/// A channel is always of some known "type", describing the transaction formats used and the exact
+/// semantics of our interaction with our peer.
+///
+/// Note that because a channel is a specific type which is proposed by the opener and accepted by
+/// the counterparty, only required features are allowed here.
```
Comment (valentinewallace) [inline lightning/src/ln/features.rs:421]: I don't think this is enforced in `from_counterparty_init` -- i think an odd `static_remotekey` bit will stay odd in the resulting `ChannelTypeFeatures` atm

**Review by valentinewallace — COMMENTED** [review]
Not 100% sure on some of these but didn't want to hold up feedback too long

File: lightning/src/ln/channel.rs [hunk 9]
```diff
@@ -5434,7 +5434,8 @@ impl<'a, Signer: Sign, K: Deref> ReadableArgs<&'a K> for Channel<Signer>
 			(11, channel_type, option),
 		});
 
-		if channel_type.as_ref().unwrap().supports_unknown_bits() || channel_type.as_ref().unwrap().requires_unknown_bits() {
+		let chan_features = channel_type.as_ref().unwrap();
+		if chan_features.supports_unknown_bits() || chan_features.requires_unknown_bits() {
```
Comment (valentinewallace) [inline lightning/src/ln/channel.rs:5438]: nit: guess a simpler check would be `if chan_features != ChannelTypeFeatures::static_key_only()`

Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:5438]: Yea, but this "feels" more right - in the future we could/will add more bits and ideally we'd not have to change this code.
